codesign/bin/codesign.dart

```diff
@@ -69,14 +69,32 @@ Future<void> main(List<String> args) async {
       help: 'Team-id is used by notary service for xcode version 13+.',
     )
     ..addOption(
-      kCommitOption,
-      help: 'the Flutter engine commit revision for which Google Cloud Storage binary artifacts should be codesigned.',
+      kGCloudDownloadPattern,
+      help: 'A required pattern to specify the google cloud bucket prefix of the artifacts stored in google cloud. '
+          'In a pattern, the word ARTIFACTRAWNAME will get replaced to the raw name of artifact without suffix or prefix, the word FILEPATH will get replaced to the artifact path\n'
+          'For example, to code sign ios usb dependency artifacts ios-deploy.zip and libplist.zip, supply the pattern\n'
+          '`--google-cloud-download-pattern = ios-usb-dependencies/unsigned/ARTIFACTRAWNAME/<commitHash>/FILEPATH`\n,'
```

Comment:
Few notes here,
1. I think this makes more sense in the README.md
2. nit: There's a bash syntax issue with space around the "="
3. I'm not sure what I'm supposed to add here. Is ARTIFACTRAWNAME expected? Is commit hash inputted? For user variables in bash, it helps to denote them as environment variables (like $ARTIFACT_RAW_NAME). Then users can copy and paste the exact command, but before it set up the inputs.
4. File path seems duplicate, and could be combined with this variable to just be `gcsPath`

Reply:
umm right i think the main confusion comes from the fact that each type of artifact requires a different type of pattern to be used. for example, ios-usb artifact might want a gcs path upload pattern of a/b/ios-usb/c/ios-usb.zip, and a download path of a/b/ios-usb/c/unsigned/ios-usb.zip, whereas an engine artifact needs a gcs path with the pattern flutter/commit/engine.zip. I think there are 3 options:

1. let user input the pattern
pros: scalable
cons: it becomes confusing for the user to craft the pattern themself.

2. assume the pattern ourselves, and let user specify the type of artifact.
pros: clear, env variables and parameters easy to use
cons: not scalable, need to add a new pattern for each new type of artifact introduced.

3. use a single gcs path parameter. I think the problem with only taking in a gcsPath parameter is that, gcs upload paths are different from gcs download paths. and when signing multiple artifacts, users might supply flags such as --upload_1 --upload_2 --download_2 --download_1, and it would be hard for code sign app to match and pair the upload and download paths.

With option 2 we can assume the pattern for users and reduce the confusions that would be caused by large blob of texts instructing users how to craft their own patterns. The downside of option 2 is that it is not as scalable as Casey pointed out, because we have to introduce a new pattern whenever a new tip of artifact is introduced. option 1 is scalable but like Casey pointed out it is confusing to read, and option 3 does not quite work because codesign app can't auto pair the download of upload paths.

umm I am thinking if there is another way of implementing it so that it is clear.

Reply:
> an engine artifact needs a gcs path with the pattern flutter/commit/engine.zip

I think you answer that in 3. For the download path, users would input gs://flutter_infra_releases/flutter/commitabc123/mac_arm/...

Can you expand on why users would need to input multiple downloads and uploads instead of rerunning the app? My understanding is the signing sub builds will only execute one download/upload.

For the issue with mapping downloads/uploads, using package:args you can have a single flag that takes in multiple values.

Reply:
Synced with Casey yesterday and the plan was to have codesign app only support signing a single artifact each time, so that we can streamline the process of generic signing, and pass in parameters without user crafting. I went ahead and removed multi artifact signing logic but there is a chance that we still need to sign multiple artifacts at the same time:

As the way it is implemented in https://flutter-review.googlesource.com/c/recipes/+/34501,
1. for global generators, they will produce multiple artifacts in the multiple out directories, and this might require a multi artifacts parameter sign.
2. for each archive, it produces multiple artifacts as shown in `included_paths`, which might require a multi parameters sign. 

To be able to sign single parameter, and align with what we would like in the sub builder design in go/code-signing-bot, my understanding is that we would need a triggering of _schedule_with_led for code signing job from within the _schedule_with_led of the sub build of the artifacts. But my understanding could be off. Would be great if sir @godofredoc could provide some insights! Thank you!

Reply:
Simpler implementations are better. Ideally we can sign a single artifact at a time and the recipe logic can take care of running the tool for as many artifacts as the build is generating.

Reply:
We can split the work here, if you tell what the interface for calling the tool is and we manually generate a cipd package I can help with the signer recipe implementation.

Reply:
Synced with sir Godofredo and my current plan is to 
1. let codesign support single parameter file path (meaning signing one artifact at most each time) 
2. parse the "parameters" field or add "included paths" in generator configs, so that we can later retrieve the total file paths to be codesigned for global generators. 
4. for sub build such as 'host_debug' which produces multiple 'included paths', or global generators, wait until the group of 'included paths' have been built, and feed them one by one to the schedule with led for codesign.